instrumentation/mongo/src/main/java/brave/mongo/TraceMongoCommandListener.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.mongo;
+
+import brave.Span;
+import brave.Tracer;
+import com.mongodb.event.CommandFailedEvent;
+import com.mongodb.event.CommandListener;
+import com.mongodb.event.CommandStartedEvent;
+import com.mongodb.event.CommandSucceededEvent;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.bson.json.JsonWriterSettings;
+
+import java.net.InetSocketAddress;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A Mongo command listener that will report via Brave how long each command takes.
+ *
+ * <p>To use it, call <code>.addCommandListener(TraceMongoCommandListener.builder().tracer(tracer).build())</code>
+ * on the {@link com.mongodb.MongoClientOptions} or {@link com.mongodb.MongoClientSettings} object that is used to
+ * create the {@code MongoClient} to be instrumented.
+ */
+public class TraceMongoCommandListener implements CommandListener {
+  private final Tracer tracer;
+  private final int maxAbbreviatedCommandLength;
+  private final JsonWriterSettings jsonWriterSettings;
+  private final Map<Integer, Span> activeSpans = new ConcurrentHashMap<>();
```

Comment:
this is a bomb waiting to happen. unbounded maps can make OOM when there's a bug that causes the hooks not to fire. Consider what we use in MySQL if there's no explicit property across start and finish.. ThreadLocalSpan

Reply:
The Mongo driver doesn't guarantee that the command listener methods will be called from the same thread, therefore I couldn't use `ThreadLocalSpan`. I'll leave a comment to that effect.

Also the `CommandListener` interface doesn't provide a mechanism for associating data with a command from start to finish... so I think a Map is pretty much the only option.

Reply:
please do put a comment but we still need to make this better before merge as the map you are using is unbounded (can grow infinitely). In worst case we need to make a bounded version of the map, or use weak references.

also can you please cite the code in mongodb-driver itself that is hopping threads?

@kojilin how are you handling this, btw.

Reply:
Currently, we are not using a command listener but just wrap the client code and delegate to the client.

Reply:
I see.. yeah this approach would prevent leaks

Reply:
> also can you please cite the code in mongodb-driver itself that is hopping threads?

@adriancole Please see [this comment](https://github.com/openzipkin/brave/pull/1090#discussion_r384055176) about threading. The gist of it is that I believe that we could get away with `ThreadLocalSpan` for the synchronous/legacy driver, although there is no guarantee that Mongo won't change the threading model in the future as I couldn't find any documented guarantees.

But `ThreadLocalSpan` is likely insufficient for the reactive/async driver as that one calls the methods from whatever IO event loop thread the callback happens to run on.

All flavors (sync/async) of the driver use the same `CommandLister` interface.

I could use `ThreadLocalSpan` for now and document that the reactive/async version is not supported. Later on I or somebody else can come back and add support for reactive in a separate pull request.

Does that sound good?

Reply:
by synchronous/legacy, you mean the 3.x series?

In the case of the reactive/async, make a failing test with a TODO perhaps? It is ok to not address cases immediately, but it helps a lot to make a failing test with `@Ignore`. I do it all the time.. ex https://github.com/spring-cloud/spring-cloud-sleuth/blob/2.2.x/spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/client/ReactorNettyHttpClientBraveTests.java#L97